File: EndGate/tests/EndGate.Core.Client.JS.Tests/Tests/UnitTests/Loopers/LooperFacts.js
(function (window, lib) {

    QUnit.module("Looper Facts");

    QUnit.asyncTimeoutTest("Looper single callback works.", testUtilities.defaultTestTimeout, function (end, assert, testName) {
        var updates = 0,
            triggered = false,
            gameLoop = new lib.Looper(),
            onComplete = function () {
                assert.equal(updates, 30, "Updates have hit 30!");
                end();
            },
            timedCallback = new lib.TimedCallback(30, function () {
                updates++;
                if (updates >= 30) {
                    triggered = true;
                    gameLoop.Dispose();
                    window.setTimeout(onComplete, 1000);
                }
            });

        gameLoop.AddCallback(timedCallback);
        gameLoop.Start();

        window.setTimeout(function () {
            if (!triggered) {
                assert.ok(false, "Loop did not finish in time!");
                end();
            }
        }, 1100);

        return function () {
            gameLoop.Dispose();
        };
    });

    QUnit.asyncTimeoutTest("Looper multiple callback works.", testUtilities.defaultTestTimeout, function (end, assert, testName) {
        var updates = 0,
            triggered = false,
            gameLoop = new lib.Looper(),
            onComplete = function () {
                assert.equal(updates, -30, "Updates have hit -30!");
                end();
            },
            timedCallback1 = new lib.TimedCallback(30, function () {
                updates++;                
            }),
            timedCallback2 = new lib.TimedCallback(60, function () {
                updates--;

                if (updates <= -30) {
                    triggered = true;
                    gameLoop.Dispose();
                    window.setTimeout(onComplete, 1000);
                }
            });

        gameLoop.AddCallback(timedCallback1);
        gameLoop.AddCallback(timedCallback2);
        gameLoop.Start();

        window.setTimeout(function () {
            if (!triggered) {
                assert.ok(false, "Loop did not finish in time!");
                end();
            }
        }, 1100);

        return function () {
            gameLoop.Dispose();
        };
    });

})(window, EndGate.Core.Loopers);

File: EndGate/tests/EndGate.Core.Net.Tests/Utilities/LooperFacts.cs
using System;
using System.Threading;
using EndGate.Core.Net.Loopers;
using Xunit;

namespace EndGate.Core.Net.Tests
{
    public class LooperFacts
    {
        [Fact]
        public void LooperSingleCallbackWorks()
        {
            long updates = 0;
            var resetEvent = new ManualResetEvent(false);
            var gameLoop = new Looper();
            var timedCallback = new TimedCallback(30, () =>
            {
                updates++;
                if (updates >= 30)
                {
                    resetEvent.Set();
                }
            });
            gameLoop.AddCallback(timedCallback);
            gameLoop.Start();

            Assert.True(resetEvent.WaitOne(TimeSpan.FromSeconds(1.01)));
            gameLoop.Dispose();

            Assert.Equal(updates, 30);
        }

        [Fact]
        public void LooperMultipleCallbacksWork()
        {
            long updates = 0;
            var resetEvent = new ManualResetEvent(false);
            var gameLoop = new Looper();
            var timedCallback1 = new TimedCallback(30, () =>
            {
                updates++;
            });
            var timedCallback2 = new TimedCallback(60, () =>
            {
                updates--;

                if (updates <= -30)
                {
                    resetEvent.Set();
                }
            });

            gameLoop.AddCallback(timedCallback1);
            gameLoop.AddCallback(timedCallback2);
            gameLoop.Start();

            Assert.True(resetEvent.WaitOne(TimeSpan.FromSeconds(1.01)));
            gameLoop.Dispose();

            // Since timedCallback2 is executing twice as fast as timedCallback1 it will negate the value of updates
            Assert.Equal(updates, -30);
        }
    }
}
